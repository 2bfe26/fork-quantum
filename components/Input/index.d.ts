import { InputHTMLAttributes, MouseEventHandler, ChangeEventHandler } from 'react';
import masks from '../shared/masks';
import Input from './Input';


import { Validations } from '../Form';

export type MaskFunction = (rawValue: string) => string[];
export type Mask = boolean | RegExp | string | MaskFunction;
export type Validate =
  | typeof Validations.Required
  | typeof Validations.CPF
  | typeof Validations.CEP
  | typeof Validations.Date
  | typeof Validations.MinLength
  | typeof Validations.MaxLength
  | typeof Validations.Email;
export type CustomValidate =
  | ((param?: { value: string }) => string)
  | { validate: Validate; error: string };

export interface InputProps extends InputHTMLAttributes<HTMLInputElement> {
  value?: string;
  label?: string;
  helperText?: string;
  descriptionLabel?: string;
  required?: boolean;
  placeholder?: string;
  type?: 'email' | 'text' | 'tel' | 'number' | 'password' | 'search';
  error?: string;
  id?: string;
  name?: string;
  validate?: Validate | CustomValidate | Array<Validate | CustomValidate>;
  mask?: Mask | Mask[];
  onClean?: MouseEventHandler<HTMLInputElement>;
  onChange?: ChangeEventHandler<HTMLInputElement>;
  minLength?: number;
  maxLength?: number;
  skin?: 'default' | 'dark';
  theme?: {
    spacing?: {};
    colors?: {};
    baseFontSize?: number;
  };
}


interface CompoundedInput extends React.FunctionComponent<InputProps> {
  CEP: React.FunctionComponent<InputProps>;
  CNPJ: React.FunctionComponent<InputProps>;
  CPF: React.FunctionComponent<InputProps>;
  Date: React.FunctionComponent<InputProps>;
  Password: React.FunctionComponent<InputProps>;
  Phone: React.FunctionComponent<InputProps>;
}

const Input: CompoundedInput = () => null

export default Input;
