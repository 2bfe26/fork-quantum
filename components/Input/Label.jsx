import styled from 'styled-components';

const InputLabel = styled.label`
  display: block;
  margin-bottom: 0;
  padding: 8px 12px 0px;
  font-weight: bold;
`;

InputLabel.displayName = 'InputLabel';

export default InputLabel;
