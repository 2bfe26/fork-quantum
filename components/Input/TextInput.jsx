import styled, { css } from 'styled-components';
import Colors from '../Colors';

const TextInput = styled.input.attrs({
  type: 'text',
})`
  box-sizing: border-box;
  margin-top: 8px;
  padding-right: 42px;
  &::-webkit-inner-spin-button,
  &::-webkit-outer-spin-button {
    -webkit-appearance: none;
    margin: 0;
  }

  background-color: ${Colors.WHITE};
  border-radius: 4px;
  border: 1.5px solid ${Colors.BLACK['400']};
  box-sizing: border-box;
  color: ${Colors.BLACK['700']};
  font-size: initial;
  height: 44px;
  min-height: 44px;
  letter-spacing: 0.2px;
  outline: none;
  padding: 10px 12px;
  transition: all 0.2s ease-in-out;
  width: 100%;

  :hover,
  :focus {
    border-color: ${Colors.BLUE['500']};
    box-shadow: 0 2px 6px 0 ${Colors.BLUE['50']};
  }

  ${({ error }) =>
    error &&
    css`
      border-color: ${Colors.ERROR['500']};

      :hover,
      :focus {
        border-color: ${Colors.ERROR['500']};
        box-shadow: 0 2px 6px 0 ${Colors.ERROR['500']};
      }
    `};

  &[disabled] {
    background-color: ${Colors.BLACK['100']};
    border-color: ${Colors.BLACK['400']};
    box-shadow: none;
    color: ${Colors.BLACK['400']};
    cursor: not-allowed;
  }

  ${({ placeholder, defaultValue, value }) =>
    placeholder &&
    !defaultValue &&
    !value &&
    `
    color: ${Colors.BLACK['400']};
  `}

  ${({ searchable }) =>
    searchable &&
    `
    padding: 10px 42px;
  `}

  ${({ password }) =>
    password &&
    `
    padding-right: 28px;
  `}

  &::-webkit-calendar-picker-indicator {
    display: none;
  }

  :-webkit-autofill {
    box-shadow: 0 0 0px 1000px ${Colors.BLUE['200']} inset;
  }
`;

TextInput.displayName = 'TextInput';

export default TextInput;
