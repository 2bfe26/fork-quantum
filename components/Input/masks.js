export default {
  cep: '99999-999',
  cnpj: '99.999.999/9999-99',
  cpf: '999.999.999-99',
  date: '99/99/9999',
  phone: '(99) 9999-99999',
};
