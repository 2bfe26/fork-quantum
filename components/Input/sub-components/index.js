import TextInput from './TextInput';
import HelperText from './HelperText';
import InputLabel from './InputLabel';
import RequiredMark from './RequiredMark';
import InputErrorMessage from './InputErrorMessage';

export { TextInput, HelperText, InputLabel, RequiredMark, InputErrorMessage };
