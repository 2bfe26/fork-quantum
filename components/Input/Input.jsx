import PropTypes from 'prop-types';
import styled from 'styled-components';
import React from 'react';
import MaskedInput from 'react-text-mask';

import { ErrorMessage, Label, FieldGroup } from '../shared';
import ColorsDeprecated from '../Colors/deprecated';
import Colors from '../Colors';
import Icon from '../Icon';
import theme from '../shared/theme';
import InputTypes from './InputTypes';

const InputLabel = styled(Label)`
  cursor: text;
  font-size: 16px;
  margin-bottom: 0;
  padding: 8px 12px 0px 12px;
  font-weight: bold;
  ${theme.mixins.transition()};
`;

const InputTag = styled.input`
  ${theme.mixins.transition()};
  background-color: transparent;
  border: none;
  border-radius: 4px;
  border: 1.5px solid ${Colors.BLACK['400']};
  box-sizing: border-box;
  color: ${Colors.BLACK['700']};
  font-size: 16px;
  height: 44px;
  padding: 10px 12px;
  outline: none;
  width: 100%;
  margin-top: 5px;

  &[disabled] {
    background-color: ${Colors.BLACK['100']};
    color: ${Colors.BLACK['400']};
    cursor: not-allowed;
    border-color: ${Colors.BLACK['400']};
    box-shadow: none;
  }

  ${({ error }) =>
    error &&
    `
    border-color: ${Colors.ERROR['500']};
  `}

  ${props =>
    props.password &&
    `
    padding-right: 28px;
  `}

  &::-webkit-calendar-picker-indicator {
    display: none;
  }

  &:focus {
    border-color: ${Colors.BLUE['500']};
    box-shadow: 0 1px 6px 0 ${Colors.BLUE['50']};
  }

  &:focus + ${InputLabel} {
    color: ${ColorsDeprecated.PRIMARY['500']};
  }

  ${({ placeholder, defaultValue }) =>
    placeholder &&
    !defaultValue &&
    `
    color: ${Colors.BLACK['400']};
  `}

  :-webkit-autofill {
    -webkit-text-fill-color: ${Colors.BLACK['700']};
    -webkit-box-shadow: 0 0 0px 1000px ${Colors.BLUE['200']} inset;
  }
`;

const InputIcon = styled(Icon)`
  cursor: pointer;
  position: absolute;
  right: 12px;
  top: 45px;
`;

const InputFieldGroup = styled(FieldGroup)`
  margin: 40px 0 20px;

  &:first-child {
    margin-top: 20px;
  }
`;

const InputErrorMessage = styled(ErrorMessage)`
  padding: 8px 12px;
`;

const HelperText = styled.span`
  color: ${Colors.BLACK['700']};
  cursor: text;
  display: block;
  font-size: 14px;
  font-style: italic;
  font-weight: 600;
  padding: 8px 12px;
`;

const DescriptionLabel = styled.span`
  color: ${Colors.BLACK['700']};
  cursor: text;
  display: block;
  font-size: 14px;
  padding: 0px 12px;
`;

const RequiredMark = styled.em`
  color: ${Colors.ERROR['500']};
`;

/** A text field component to get user text data */
class Input extends React.Component {
  static CEP = InputTypes.CEP;

  static CNPJ = InputTypes.CNPJ;

  static CPF = InputTypes.CPF;

  static Date = InputTypes.Date;

  static Phone = InputTypes.Phone;

  static Password = InputTypes.Password;

  constructor(props) {
    super(props);

    const { value, type } = props;

    this.state = {
      value,
      type,
    };
  }

  componentWillUpdate(nextProps) {
    const { value, type } = this.state;
    if (nextProps.value !== value || nextProps.type !== type) {
      this.state.value = nextProps.value;
      this.state.type = nextProps.type;
    }
  }

  _onChange = e => {
    const { onChange } = this.props;
    const {
      target: { value },
    } = e;

    this.setState({ value });

    onChange(e, { value });
  };

  _changeType = type => {
    this.setState({ type });
  };

  _showPassword = () => {
    const { type } = this.state;

    if (type === 'text') {
      this._changeType('password');
    } else {
      this._changeType('text');
    }
  };

  _handleClear = () => {
    this.setState({ value: '' });
  };

  render() {
    const {
      id,
      label,
      error,
      mask,
      type: inputType,
      descriptionLabel,
      helperText,
      required,
      ...rest
    } = this.props;
    const { value, type } = this.state;
    const valueIsTyped = !!value;

    return (
      <InputFieldGroup>
        {label && (
          <InputLabel htmlFor={id} error={error}>
            {label}
            {required && <RequiredMark>*</RequiredMark>}
          </InputLabel>
        )}
        {descriptionLabel && (
          <DescriptionLabel>{descriptionLabel}</DescriptionLabel>
        )}
        <MaskedInput
          {...rest}
          id={id}
          type={type}
          mask={mask}
          value={value}
          onChange={this._onChange}
          render={(ref, props) => (
            <InputTag ref={ref} error={error} {...props} />
          )}
        />
        {inputType === 'password' && (
          <InputIcon
            name={type === 'password' ? 'visibility' : 'visibility_off'}
            onClick={this._showPassword}
          />
        )}
        {valueIsTyped && (
          <InputIcon name="cancel" onClick={this._handleClear} />
        )}
        {error && <InputErrorMessage>{error}</InputErrorMessage>}
        {helperText && !error && <HelperText>{helperText}</HelperText>}
      </InputFieldGroup>
    );
  }
}

Input.defaultProps = {
  error: '',
  id: '',
  label: '',
  mask: false,
  maxLength: '',
  type: 'text',
  value: '',
  helperText: '',
  descriptionLabel: '',
  required: false,
  disabled: false,
  placeholder: '',
  onBlur: () => {},
  onChange: () => {},
  onFocus: () => {},
};

Input.propTypes = {
  value: PropTypes.string,
  /** Display a label text that describe the field */
  label: PropTypes.string,
  /** Display a helper text below the input */
  helperText: PropTypes.string,
  /** Display a description text below the label */
  descriptionLabel: PropTypes.string,
  /** set if the input is required */
  required: PropTypes.bool,
  disabled: PropTypes.bool,
  placeholder: PropTypes.string,
  type: PropTypes.oneOf(['email', 'text', 'tel', 'number', 'password']),
  /** Display an error message and changes border color to error color */
  error: PropTypes.string,
  /** Set a text mask that filter user input */
  maxLength: PropTypes.string,
  /** A html identification */
  id: PropTypes.string,
  /**
   * Mask must follow this [rules](https://github.com/text-mask/text-mask/blob/master/componentDocumentation.md#mask)
   */
  mask: PropTypes.oneOfType([
    PropTypes.array,
    PropTypes.bool,
    PropTypes.instanceOf(RegExp),
    PropTypes.func,
    PropTypes.string,
  ]),
  onBlur: PropTypes.func,
  onChange: PropTypes.func,
  onFocus: PropTypes.func,
};

Input.displayName = 'Input';

export default Input;
