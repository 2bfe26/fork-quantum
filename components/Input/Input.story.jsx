
import React from 'react';
import { storiesOf } from '@storybook/react';
import Heading from '../../.storybook/components/Heading';
import AutoPropsApi from '../../.storybook/components/AutoPropsApi';
import ComponentPanel from '../../.storybook/components/ComponentPanel';
import Example from '../../.storybook/components/Example';
import { TabbedView, Tab } from '../../.storybook/components/TabbedView';
import Input from './Input';

storiesOf('3. Forms', module)
  .add('Input', () => (
    <Heading atom title="Input">
      <TabbedView>
        <Tab title="Usage">
          <ComponentPanel component={<Input />} importModules="Input" />
        </Tab>

        <Tab title="Example">
          <Example title="CEP" component={<Input.CEP style={{ width: '200px' }} placeholder="CEP" />} importModules="Input" />
          <Example title="CNPJ" component={<Input.CNPJ style={{ width: '200px' }} />} importModules="Input" />
          <Example title="CPF" component={<Input.CPF style={{ width: '200px' }} />} importModules="Input" />
          <Example title="Date" component={<Input.Date style={{ width: '200px' }} />} importModules="Input" />
          <Example title="Phone" component={<Input.Phone style={{ width: '200px' }} />} importModules="Input" />
        </Tab>

        <Tab title="API">
          <AutoPropsApi component={Input} />
        </Tab>
      </TabbedView>
      <Input.CEP />
    </Heading>
  ));
