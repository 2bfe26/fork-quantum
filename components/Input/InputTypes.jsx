import React from 'react';
import Input from './Input';
import masks from '../shared/masks';

const types = {
  CEP: props => (<Input mask={masks.cep} {...props} />),
  CNPJ: props => (<Input mask={masks.cnpj} {...props} />),
  CPF: props => (<Input mask={masks.cpf} {...props} />),
  Date: props => (<Input mask={masks.date} {...props} />),
  Phone: props => (<Input mask={masks.phone} {...props} />),
  Password: props => (<Input type="password" password {...props} />),
};

types.CEP.displayName = 'Input.CEP';
types.CNPJ.displayName = 'Input.CNPJ';
types.CPF.displayName = 'Input.CPF';
types.Date.displayName = 'Input.Date';
types.Phone.displayName = 'Input.Phone';

export default types;
