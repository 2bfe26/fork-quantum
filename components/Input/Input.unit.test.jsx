import React from 'react';
import renderer from 'react-test-renderer';
import { shallow, mount } from 'enzyme';
import MaskedInput from 'react-text-mask';
import Input from './Input';
import masks from '../shared/masks';

describe('Input component ', () => {
  it('should match snapshots', () => {
    const INPUTS = [
      <Input />,
      <Input value="foo" />,
      <Input label="Text label" />,
      <Input error="Error message" />,
      <Input required />,
      <Input searchable />,
      <Input disabled />,
      <Input helperText="this is a helper text" />,
      <Input descriptionLabel="this is a description label" />,
      <Input descriptionLabel="this is a description label" searchable />,
      <Input placeholder="this input has a placeholder" />,
    ];

    INPUTS.forEach(input =>
      expect(renderer.create(input).toJSON()).toMatchSnapshot(),
    );
  });

  it('should has a required signal when "required" prop is set', () => {
    const component = shallow(
      <Input value="foo" label="label of input" required />,
    );
    const label = component.childAt(0);
    const requiredElement = label
      .childAt(1)
      .html()
      .includes('*');

    expect(requiredElement).toBeTruthy();
  });

  it('should has a searchable icon when "searchable" prop is set', () => {
    const component = shallow(
      <Input value="foo" label="label of input" searchable />,
    );
    const InputSearchIconElement = component
      .find('InputSearchIcon')
      .html()
      .includes('search');

    expect(InputSearchIconElement).toBeTruthy();
  });

  it('should has an error icon when "error" prop is set', () => {
    const component = shallow(
      <Input value="foo" label="label of input" error="error text" />,
    );
    const InputErrorIconElement = component
      .find('InputErrorIcon')
      .html()
      .includes('error');

    expect(InputErrorIconElement).toBeTruthy();
  });

  it('should has a helper text when "helperText" prop is set ', () => {
    const helperTextContent = 'this is a helper text';
    const component = shallow(
      <Input
        value="foo"
        label="label of input"
        helperText={helperTextContent}
      />,
    );
    const helperTextElementContent = component.find('HelperText').text();

    expect(helperTextElementContent).toEqual(helperTextContent);
  });

  it('should has a description label when "descriptionLabel" prop was set', () => {
    const descriptionLabelContent = 'this is a description label';
    const component = shallow(
      <Input
        value="foo"
        label="label of input"
        descriptionLabel={descriptionLabelContent}
      />,
    );
    const descriptionLabelElementContent = component
      .find('DescriptionLabel')
      .text();

    expect(descriptionLabelElementContent).toEqual(descriptionLabelContent);
  });

  it('should has a placeholder when "placeholder" is text', () => {
    const placeholderContent = 'this is a input placeholder';
    const component = mount(
      <Input
        value="foo"
        label="label of input"
        placeholder={placeholderContent}
      />,
    );

    const placeholderProp = component.find('InputTag').prop('placeholder');
    expect(placeholderProp).toEqual(placeholderContent);
  });

  it('should has a cancel icon when input value is not empty', () => {
    const component = mount(<Input value="foo" label="label of input" />);
    const hasIcon = component.find('span.material-icons');
    expect(hasIcon).toBeTruthy();

    const cancelIcon = hasIcon.text();
    expect(cancelIcon).toEqual('cancel');
  });

  it('should generate a new id for it instance, when "id" prop is not set', () => {
    const componentA = shallow(<Input value="foo" label="label of input" />);
    const componentB = shallow(<Input value="foo" label="label of input" />);
    const inputIdA = componentA.find(MaskedInput).prop('id');
    const inputIdB = componentB.find(MaskedInput).prop('id');

    expect(inputIdA).not.toBe('');
    expect(inputIdB).not.toBe('');

    expect(inputIdA).not.toEqual(inputIdB);
  });

  it('should apply a mask prop to input value', () => {
    const input = mount(<Input mask={masks.cpf} value="99999999999" />);
    const { value } = input.find(MaskedInput).getDOMNode();

    expect(value).toBe('999.999.999-99');
  });

  describe('with a label', () => {
    it('should match label "htmlFor" label param with "id" input param', () => {
      const input = <Input label="Text label" id="input-id" value="foo" />;
      const wrapper = mount(input);
      const inputTag = wrapper.find('InputTag');
      const label = wrapper.find('InputLabel');

      expect(label.prop('htmlFor')).toEqual(inputTag.prop('id'));
    });

    it('should match label "htmlFor" label param and "input" param with generated id', () => {
      const wrapper = shallow(<Input label="Text label" value="foo" />);
      const labelId = wrapper.find('InputLabel').prop('htmlFor');
      const inputId = wrapper.find(MaskedInput).prop('id');

      expect(labelId).toEqual(inputId);
    });
  });

  describe('with password type', () => {
    const wrapper = shallow(<Input type="password" />);

    const icon = () => wrapper.childAt(1);

    const visibilityIcon = () =>
      icon()
        .html()
        .includes('visibility');

    it('should has "password" as default input type', () => {
      expect(visibilityIcon()).toBeTruthy();
      expect(wrapper.state('type')).toEqual('password');
    });

    it('should toggle input type, when password icon is clicked', () => {
      const visibilityOffIcon = () =>
        icon()
          .html()
          .includes('visibility_off');

      icon().simulate('click');
      expect(visibilityOffIcon()).toBeTruthy();
      expect(wrapper.state('type')).toEqual('text');

      icon().simulate('click');
      expect(visibilityIcon()).toBeTruthy();
      expect(wrapper.state('type')).toEqual('password');
    });
  });
});
