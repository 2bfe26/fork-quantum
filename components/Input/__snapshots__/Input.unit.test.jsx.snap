// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input component  should match the snapshot 1`] = `
<div
  className="sc-htpNat QOKog"
>
  
  <input
    className="sc-bxivhb cDkyWh"
    defaultValue="foo"
    id=""
    maxLength=""
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    placeholder=""
    type="text"
  />
  
</div>
`;

exports[`Input component  with a label should match the snapshot 1`] = `
<div
  className="sc-htpNat QOKog"
>
  <label
    className="sc-bwzfXH eQhPVJ"
    htmlFor="input-with-label"
  >
     
    Text label
     
  </label>
  <input
    className="sc-bxivhb cDkyWh"
    defaultValue="foo"
    id="input-with-label"
    maxLength=""
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    placeholder=""
    type="text"
  />
  
</div>
`;

exports[`Input component  with an error message should match the snapshot 1`] = `
<div
  className="sc-htpNat QOKog"
>
  
  <input
    className="sc-bxivhb ezFnNS"
    defaultValue="foo"
    id=""
    maxLength=""
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    placeholder=""
    type="text"
  />
  <div
    className="sc-bdVaJa fIlZBq"
  >
    Error message
  </div>
</div>
`;
